File: labs/23-code-review-agent/node/main.ts
import fs from "fs";
import path from "path";
import { spawnSync } from "child_process";
import { OpenAIClient } from "../../../shared/config/openai_client";

function findRuff(): string | null {
  const candidates = ["ruff", "ruff.exe"];
  for (const bin of candidates) {
    const result = spawnSync("which", [bin], { encoding: "utf8" });
    if (result.status === 0 && result.stdout.trim()) return result.stdout.trim();
  }
  return null;
}

function runSyntaxCheck(filePath: string): { ok: boolean; issues: string[] } {
  const result = spawnSync("python", ["-m", "py_compile", filePath], {
    encoding: "utf8",
  });
  if (result.status === 0) return { ok: true, issues: [] };
  return { ok: false, issues: [`Syntax error: ${result.stderr || result.stdout}`] };
}

function runRuff(filePath: string): { ok: boolean; issues: string[] } {
  const ruffBin = findRuff();
  if (!ruffBin) {
    console.warn("ruff not installed; skipping style check");
    return { ok: true, issues: ["ruff not installed"] };
  }
  const result = spawnSync(ruffBin, ["check", filePath, "--output-format", "json"], {
    encoding: "utf8",
  });
  const issues: string[] = [];
  if (result.stdout.trim()) {
    try {
      const parsed = JSON.parse(result.stdout);
      for (const item of parsed) {
        issues.push(
          `Line ${item.location?.row}: ${item.code} - ${item.message}`
        );
      }
    } catch {
      issues.push(result.stdout.trim());
    }
  }
  return { ok: issues.length === 0, issues };
}

async function llmReview(
  filePath: string,
  source: string,
  checks: any,
  client: OpenAIClient | null
): Promise<string> {
  if (!client) return "LLM review skipped (no API key). Manual review recommended.";

  const prompt = `Review the following Python file for style, bugs, and maintainability.

File: ${path.basename(filePath)}

Static checks:
- Syntax OK: ${checks.syntax.ok}
- Style issues: ${checks.ruff.issues}

Source code:
\`\`\`python
${source}
\`\`\`

Provide a concise review with:
1. Critical issues
2. Suggestions
3. Overall verdict (OK / Needs work).`;

  const resp = await client.chatCompletion({
    messages: [{ role: "user", content: prompt }],
    temperature: 0.2,
    max_tokens: 400,
  });
  return resp.choices[0].message.content.trim();
}

async function reviewFile(filePath: string, client: OpenAIClient | null): Promise<any> {
  console.log(`Reviewing ${filePath}`);
  const source = fs.readFileSync(filePath, "utf8");
  const checks = {
    syntax: runSyntaxCheck(filePath),
    ruff: runRuff(filePath),
  };
  const review = await llmReview(filePath, source, checks, client);
  const ruffIssues = checks.ruff.issues.filter((i) => i !== "ruff not installed");
  return {
    file: filePath,
    checks,
    llm_review: review,
    verdict:
      checks.syntax.ok && ruffIssues.length === 0 ? "ok" : "needs_work",
  };
}

async function main() {
  let client: OpenAIClient | null = null;
  try {
    client = new OpenAIClient();
  } catch (exc: any) {
    console.warn("LLM client disabled:", exc.message);
  }

  const target = process.argv[2]
    ? path.resolve(process.argv[2])
    : path.resolve("sample_code.py");

  if (!fs.existsSync(target)) {
    console.error("File not found:", target);
    process.exit(1);
  }

  const report = await reviewFile(target, client);
  console.log(JSON.stringify(report, null, 2));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: labs/23-code-review-agent/python/main.py
"""Code review agent pipeline.

Reads a Python file, runs static/style checks, sends to an LLM for review,
and aggregates a report.
"""
import argparse
import json
import logging
import py_compile
import shutil
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import OpenAIClient, configure_logging

logger = logging.getLogger("code-review-agent")


def _find_ruff() -> str | None:
    if ruff := shutil.which("ruff"):
        return ruff
    candidate = Path(sys.executable).with_name("ruff")
    if candidate.exists():
        return str(candidate)
    candidate_exe = Path(sys.executable).with_name("ruff.exe")
    if candidate_exe.exists():
        return str(candidate_exe)
    return None


def run_syntax_check(path: Path) -> dict:
    try:
        py_compile.compile(str(path), doraise=True)
        return {"ok": True, "issues": []}
    except py_compile.PyCompileError as exc:
        return {"ok": False, "issues": [f"Syntax error: {exc}"]}


def run_ruff(path: Path) -> dict:
    ruff_bin = _find_ruff()
    if not ruff_bin:
        logger.warning("ruff not installed; skipping style check")
        return {"ok": True, "issues": ["ruff not installed"]}
    try:
        result = subprocess.run(
            [ruff_bin, "check", str(path), "--output-format", "json"],
            capture_output=True,
            text=True,
            check=False,
        )
        issues = []
        if result.stdout.strip():
            try:
                parsed = json.loads(result.stdout)
                for item in parsed:
                    location = item.get("location", {})
                    issues.append(
                        f"Line {location.get('row')}: {item.get('code')} - {item.get('message')}"
                    )
            except json.JSONDecodeError:
                issues.append(result.stdout.strip())
        return {"ok": len(issues) == 0, "issues": issues}
    except FileNotFoundError:
        logger.warning("ruff not installed; skipping style check")
        return {"ok": True, "issues": ["ruff not installed"]}


def llm_review(path: Path, source: str, checks: dict, client: OpenAIClient | None) -> str:
    if not client:
        return "LLM review skipped (no API key). Manual review recommended."

    prompt = f"""Review the following Python file for style, bugs, and maintainability.

File: {path.name}

Static checks:
- Syntax OK: {checks['syntax']['ok']}
- Style issues: {checks['ruff']['issues']}

Source code:
```python
{source}
```

Provide a concise review with:
1. Critical issues
2. Suggestions
3. Overall verdict (OK / Needs work)."""

    resp = client.chat_completion(
        messages=[{"role": "user", "content": prompt}],
        temperature=0.2,
        max_tokens=400,
    )
    return resp["choices"][0]["message"]["content"].strip()


def review_file(path: Path, client: OpenAIClient | None) -> dict:
    logger.info("Reviewing %s", path)
    source = path.read_text(encoding="utf-8")
    checks = {
        "syntax": run_syntax_check(path),
        "ruff": run_ruff(path),
    }
    review = llm_review(path, source, checks, client)
    ruff_issues = [i for i in checks["ruff"]["issues"] if i != "ruff not installed"]
    return {
        "file": str(path),
        "checks": checks,
        "llm_review": review,
        "verdict": "needs_work" if not checks["syntax"]["ok"] or ruff_issues else "ok",
    }


def main() -> None:
    configure_logging()
    parser = argparse.ArgumentParser(description="Code review agent")
    parser.add_argument(
        "file", nargs="?", default="sample_code.py", help="Python file to review"
    )
    args = parser.parse_args()

    client = None
    try:
        client = OpenAIClient()
    except ValueError as exc:
        logger.warning("LLM client disabled: %s", exc)

    target = Path(args.file)
    if not target.exists():
        logger.error("File not found: %s", target)
        sys.exit(1)

    report = review_file(target, client)
    print(json.dumps(report, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()
